# 1. Introduction

For this project i will be working with a dataset of used cars from ebay kleinanzeigen.

The original dataset has been scraped and uploaded to kaggle. A few modifications have been made:
1. 50,000 data points were sampled from the full dataset to ensure the codr runs quickly.
2. The dataset was dirtied to more closely resemble what would be expected from a scraped dataset.



The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.
name - Name of the car.
seller - Whether the seller is private or a dealer.
offerType - The type of listing
price - The price on the ad to sell the car.
abtest - Whether the listing is included in an A/B test.
vehicleType - The vehicle Type.
yearOfRegistration - The year in which the car was first registered.
gearbox - The transmission type.
powerPS - The power of the car in PS.
model - The car model name.
kilometer - How many kilometers the car has driven.
monthOfRegistration - The month in which the car was first registered.
fuelType - What type of fuel the car uses.
brand - The brand of the car.
notRepairedDamage - If the car has a damage which is not yet repaired.
dateCreated - The date on which the eBay listing was created.
nrOfPictures - The number of pictures in the ad.
postalCode - The postal code for the location of the vehicle.
lastSeenOnline - When the crawler saw this ad last online.

Goal: Clean the data and analyze the included used car listings
Objective: Learn to use pandas and numpy tools within jupyter notebook.

In [1]:
import numpy as np
import pandas as pd

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:

autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Using describe we can see that there are a few columns null values:
1. vehicleType
2. gearbox
3. model
4. fuelType
5. notRepairedDamage

5 columns have ints and the rest are strings or mixed.
The columns names use camelcase.

# 2. Cleaning Column Names

Now we want to convert the column names from camelcase to snakecase and reword some of the columns names based on the data dictionary to be more descriptive.

In [4]:
col_names = autos.columns
new_names = ["date_crawled","name","seller","offer_type","price",
            "abtest", "vehicle_type", "registration_year", "gearbox",
            "power_ps","model", "odometer", "registration_month","fuel_type",
            "brand", "unrepaired_damage", "ad_created", "nr_of_pictures",
            "postal_code", "last_seen"]
col_map = {}
for i in range(0,len(col_names)):
    col_map[col_names[i]] = new_names[i]
    


In [5]:
autos.columns = col_names.map(col_map)
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

to make it easier to read and use we changed the column names to snakecase and reduced the words that are contain in certain column names for brevity.

# 3. Initial Exploration and Cleaning

Now we will do some more data exploration to determine what other cleaning tasks need to be done.

Things to check for:
1. Text columns where almost all thevalues are the same. if there is little or no variability the information isn't particularly useful.
2. Examples of numeric data stored as text which can be cleaned and converted


In [7]:
autos.describe()

registration_year      power_ps  registration_month  nr_of_pictures  \
count       50000.000000  50000.000000        50000.000000         50000.0   
mean         2005.073280    116.355920            5.723360             0.0   
std           105.712813    209.216627            3.711984             0.0   
min          1000.000000      0.000000            0.000000             0.0   
25%          1999.000000     70.000000            3.000000             0.0   
50%          2003.000000    105.000000            6.000000             0.0   
75%          2008.000000    150.000000            9.000000             0.0   
max          9999.000000  17700.000000           12.000000             0.0   

        postal_code  
count  50000.000000  
mean   50813.627300  
std    25779.747957  
min     1067.000000  
25%    30451.000000  
50%    49577.000000  
75%    71540.000000  
max    99998.000000

nr_of_pictures column seems to have only 1 value

registration_month happens to have a 0 as a min which means that at least one of the cars has 0 as a value for that column. That doesn't really correspond to any month so we need to deal with that.

registration_year happens to have a min of 1000 and a max of 9999 whic don't really makes sense because cars have only been around for less than 200 years and we are only in year 2020. so registering a car in the future doesn't make sense. We're gonna have to establish a time period for what range or registration years are valid.

The odometer and price have some text mixed with numeric values so we're going to have clean those columns so we'll start of by cleaning those since they're the simplest.

In [8]:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(int)
autos = autos.rename(columns={"price":"price_dollars",
                      "odometer":"odometer_km"
                     })


In [9]:
autos.describe()

price_dollars  registration_year      power_ps    odometer_km  \
count   5.000000e+04       50000.000000  50000.000000   50000.000000   
mean    9.840044e+03        2005.073280    116.355920  125732.700000   
std     4.811044e+05         105.712813    209.216627   40042.211706   
min     0.000000e+00        1000.000000      0.000000    5000.000000   
25%     1.100000e+03        1999.000000     70.000000  125000.000000   
50%     2.950000e+03        2003.000000    105.000000  150000.000000   
75%     7.200000e+03        2008.000000    150.000000  150000.000000   
max     1.000000e+08        9999.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        50000.000000         50000.0  50000.000000  
mean             5.723360             0.0  50813.627300  
std              3.711984             0.0  25779.747957  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30451.000000  
50%              6.000000             0.0  49577.000000  
75%              9.000000             0.0  71540.000000  
max             12.000000             0.0  99998.000000

# 4. Exploring the Odometer and Price Columns

We're gonna continue exploring the odometer_km an price_dollars columns. We're gonna take the following steps:
1. analyze the columns using min and max values and look for any values that look like outliers so we can remove them.
2. Then remove those specific values.

In [10]:
autos["odometer_km"].unique().shape

(13,)

In [11]:
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [12]:
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
40000       819
30000       789
20000       784
10000       264
5000        967
Name: odometer_km, dtype: int64

In [13]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        967
10000       264
20000       784
30000       789
40000       819
50000      1027
60000      1164
70000      1230
80000      1436
90000      1757
100000     2169
125000     5170
150000    32424
Name: odometer_km, dtype: int64

In [14]:
autos["price_dollars"].unique().shape

(2357,)

In [15]:
autos["price_dollars"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

In [16]:
autos["price_dollars"].value_counts().sort_index(ascending=False).head()

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
Name: price_dollars, dtype: int64

In [17]:
autos["price_dollars"].value_counts().sort_index(ascending=True).head()

0    1421
1     156
2       3
3       1
5       2
Name: price_dollars, dtype: int64

In [18]:
autos = autos[autos["price_dollars"].between(1,27322222)]

In [19]:
autos["price_dollars"].value_counts().sort_index(ascending=True).head(10)

1     156
2       3
3       1
5       2
8       1
9       1
10      7
11      2
12      3
13      2
Name: price_dollars, dtype: int64

# 5. Exploring the Date Columns

Now we're going to analyze any columns that have dates. There ar 5 columns:
1. `date_crawled`: added by the crawler
2. `last_seen`: added by the crawler
3. `ad_created`: from the website
4. `registration_month`: from the website
5. `registration_year`: from the website

Right now 1, 2, and 3 are identfied as strings so we're going to need to convert them into numerical representations in order to understand the data quantitatively

We need to see how the values for those string columns are formatted.

In [20]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [21]:
autos_date_crawled = autos['date_crawled'].str[:10]
autos_date_crawled.value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025320
2016-03-06    0.014039
2016-03-07    0.036004
2016-03-08    0.033328
2016-03-09    0.033101
2016-03-10    0.032175
2016-03-11    0.032566
2016-03-12    0.036930
2016-03-13    0.015666
2016-03-14    0.036539
2016-03-15    0.034275
2016-03-16    0.029602
2016-03-17    0.031640
2016-03-18    0.012907
2016-03-19    0.034769
2016-03-20    0.037877
2016-03-21    0.037404
2016-03-22    0.032998
2016-03-23    0.032216
2016-03-24    0.029334
2016-03-25    0.031599
2016-03-26    0.032196
2016-03-27    0.031084
2016-03-28    0.034851
2016-03-29    0.034131
2016-03-30    0.033678
2016-03-31    0.031846
2016-04-01    0.033678
2016-04-02    0.035469
2016-04-03    0.038598
2016-04-04    0.036519
2016-04-05    0.013092
2016-04-06    0.003170
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The date crawled spans a month from march to april

In [22]:
autos_ad_created = autos['ad_created'].str[:10]
autos_ad_created.value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033163
2016-03-10    0.031887
2016-03-11    0.032896
2016-03-12    0.036766
2016-03-13    0.017004
2016-03-14    0.035181
2016-03-15    0.034007
2016-03-16    0.030117
2016-03-17    0.031290
2016-03-18    0.013586
2016-03-19    0.033678
2016-03-20    0.037939
2016-03-21 

Most of the ads were created after february 2016 and there is a drop off in april.

In [23]:
autos_last_seen = autos['last_seen'].str[:10]
autos_last_seen.value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001070
2016-03-06    0.004323
2016-03-07    0.005393
2016-03-08    0.007431
2016-03-09    0.009613
2016-03-10    0.010663
2016-03-11    0.012372
2016-03-12    0.023797
2016-03-13    0.008893
2016-03-14    0.012598
2016-03-15    0.015871
2016-03-16    0.016448
2016-03-17    0.028079
2016-03-18    0.007349
2016-03-19    0.015830
2016-03-20    0.020647
2016-03-21    0.020647
2016-03-22    0.021368
2016-03-23    0.018527
2016-03-24    0.019762
2016-03-25    0.019206
2016-03-26    0.016798
2016-03-27    0.015645
2016-03-28    0.020874
2016-03-29    0.022356
2016-03-30    0.024764
2016-03-31    0.023797
2016-04-01    0.022788
2016-04-02    0.024929
2016-04-03    0.025197
2016-04-04    0.024476
2016-04-05    0.124768
2016-04-06    0.221808
2016-04-07    0.131912
Name: last_seen, dtype: float64

The last seen dates span amount from march to april of a single year

In [24]:
autos["registration_year"].describe()

count    48578.000000
mean      2004.753119
std         88.632571
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

As mentioned earlier 1000 and 9999 are values that don't make sense for year because cars weren't around back then and 9999 is the future.

# 6. Dealing with Incorrect Registration Year Data

Because these are listings for 2016 the we can choose that the only valid registration years are between 1900-2016.

We need to test this first so we'll include them and then exclude them to see if it makes a difference.


In [25]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=True)

1000    0.000021
1001    0.000021
1111    0.000021
1800    0.000041
1910    0.000103
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000041
1937    0.000082
1938    0.000021
1939    0.000021
1941    0.000041
1943    0.000021
1948    0.000021
1950    0.000062
1951    0.000041
1952    0.000021
1953    0.000021
1954    0.000041
1955    0.000041
1956    0.000082
1957    0.000041
1958    0.000082
1959    0.000124
1960    0.000494
1961    0.000124
1962    0.000082
1963    0.000165
1964    0.000247
          ...   
2000    0.064968
2001    0.054284
2002    0.051175
2003    0.055560
2004    0.055642
2005    0.060439
2006    0.054984
2007    0.046873
2008    0.045597
2009    0.042941
2010    0.032710
2011    0.033410
2012    0.026967
2013    0.016530
2014    0.013669
2015    0.008069
2016    0.025114
2017    0.028676
2018    0.009696
2019    0.000041
2800    0.000021
4100    0.000021
4500    0.000021
4800    0.000021
5000    0.000082
5911    0.000021
6200    0.000021
8888    0.0000

In [26]:
reg_year_bool = (autos["registration_year"] >=1900) &(autos["registration_year"]<=2016)

In [27]:
autos = autos[reg_year_bool] 

In [28]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=True)

1910    0.000107
1927    0.000021
1929    0.000021
1931    0.000021
1934    0.000043
1937    0.000086
1938    0.000021
1939    0.000021
1941    0.000043
1943    0.000021
1948    0.000021
1950    0.000064
1951    0.000043
1952    0.000021
1953    0.000021
1954    0.000043
1955    0.000043
1956    0.000086
1957    0.000043
1958    0.000086
1959    0.000129
1960    0.000514
1961    0.000129
1962    0.000086
1963    0.000171
1964    0.000257
1965    0.000364
1966    0.000471
1967    0.000557
1968    0.000557
          ...   
1987    0.001542
1988    0.002891
1989    0.003727
1990    0.007432
1991    0.007260
1992    0.007946
1993    0.009102
1994    0.013471
1995    0.026300
1996    0.029405
1997    0.041784
1998    0.050608
1999    0.062066
2000    0.067592
2001    0.056476
2002    0.053243
2003    0.057804
2004    0.057890
2005    0.062880
2006    0.057205
2007    0.048766
2008    0.047439
2009    0.044676
2010    0.034032
2011    0.034760
2012    0.028056
2013    0.017198
2014    0.0142

We can ask a few more questions about what's plausible. Old vintage cars would be quite expensive so we need to see if the prices for any 100 year old cars matches.

In [29]:
autos[(autos["registration_year"] == 1910)]

date_crawled                              name  seller  \
3679   2016-04-04 00:36:17                        Suche_Auto  privat   
22659  2016-03-14 08:51:18                      Opel_Corsa_B  privat   
28693  2016-03-22 17:48:41                    Renault_Twingo  privat   
30781  2016-03-25 13:47:46  Opel_Calibra_V6_DTM_Bausatz_1:24  privat   
45157  2016-03-11 22:37:01                        Motorhaube  privat   

      offer_type  price_dollars   abtest vehicle_type  registration_year  \
3679     Angebot              1     test          NaN               1910   
22659    Angebot            500     test          NaN               1910   
28693    Angebot            599  control   kleinwagen               1910   
30781    Angebot             30     test          NaN               1910   
45157    Angebot             15  control          NaN               1910   

       gearbox  power_ps    model  odometer_km  registration_month fuel_type  \
3679       NaN         0      NaN         5000                   0       NaN   
22659      NaN         0    corsa       150000                   0       NaN   
28693  manuell         0      NaN         5000                   0    benzin   
30781      NaN         0  calibra       100000                   0       NaN   
45157      NaN         0      NaN         5000                   0       NaN   

                brand unrepaired_damage           ad_created  nr_of_pictures  \
3679   sonstige_autos               NaN  2016-04-04 00:00:00               0   
22659            opel               NaN  2016-03-14 00:00:00               0   
28693         renault               NaN  2016-03-22 00:00:00               0   
30781            opel               NaN  2016-03-25 00:00:00               0   
45157         trabant               NaN  2016-03-11 00:00:00               0   

       postal_code            last_seen  
3679         40239  2016-04-04 07:49:15  
22659        52393  2016-04-03 07:53:55  
28693        70376  2016-04-06 09:16:59  
30781        47638  2016-03-26 23:46:29  
45157        90491  2016-03-25 11:18:57

We're going to remove the cars for 1910 because the prices don't seem feasable.

In [30]:
reg_year_bool = (autos["registration_year"] >=1910) &(autos["registration_year"]<=2016)
autos = autos[reg_year_bool]

# 7. Exploring Price by Brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

In [31]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [32]:
autos["brand"].value_counts(normalize=True).sort_values(ascending=False)

volkswagen        0.211278
bmw               0.110040
opel              0.107577
mercedes_benz     0.096441
audi              0.086546
ford              0.069926
renault           0.047139
peugeot           0.029834
fiat              0.025636
seat              0.018269
skoda             0.016405
nissan            0.015270
mazda             0.015185
smart             0.014157
citroen           0.014028
toyota            0.012700
hyundai           0.010023
sonstige_autos    0.009873
volvo             0.009145
mini              0.008760
mitsubishi        0.008224
honda             0.007839
kia               0.007068
alfa_romeo        0.006639
porsche           0.006125
suzuki            0.005932
chevrolet         0.005697
chrysler          0.003512
dacia             0.002634
daihatsu          0.002506
jeep              0.002270
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001563
daewoo            0.001499
trabant           0.001392
r

We'll aggregate in two ways first by the top 20 and then by any brand contributing 5% or more.

In [33]:
top_20_brands = autos["brand"].value_counts().sort_values(ascending=False).head(20).index

In [34]:
aggregate_data = {}
for b in top_20_brands: 
    selected_rows = autos[autos["brand"] == b]
    mean_price = selected_rows["price_dollars"].mean() 
    aggregate_data[b] = mean_price

In [35]:
aggregate_data

{'volkswagen': 6729.81956411556,
 'bmw': 8571.480147917478,
 'opel': 5432.479195699781,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 7456.547932618683,
 'renault': 2474.8646069968195,
 'peugeot': 3094.0172290021537,
 'fiat': 2813.748538011696,
 'seat': 4397.230949589683,
 'skoda': 6368.0,
 'nissan': 4743.40252454418,
 'mazda': 4112.596614950635,
 'smart': 3580.2239031770046,
 'citroen': 45486.68549618321,
 'toyota': 5167.091062394604,
 'hyundai': 5365.254273504273,
 'sonstige_autos': 45208.36442516269,
 'volvo': 4946.501170960188,
 'mini': 10613.459657701711}

In [36]:
brands_5_percent = autos["brand"].value_counts(normalize=True).sort_values(ascending=False).head(6).index

In [37]:
aggregate_data_2 = {}
for b in brands_5_percent: 
    selected_rows = autos[autos["brand"] == b]
    mean_price = selected_rows["price_dollars"].mean() 
    aggregate_data_2[b] = mean_price

In [38]:
aggregate_data_2

{'volkswagen': 6729.81956411556,
 'bmw': 8571.480147917478,
 'opel': 5432.479195699781,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 7456.547932618683}

Comparing the 2 ways we aggregated the data based on brand we can see that on average the mean price is a bit higher for both the top brands. one of the brands not contributing at least top 5% of sales was citreon sonstige_autos and mini which had means that were quite high. Higher than the rest of the brands.

# 8. Storing aggregate data in a dataframe

We are now going take aggregate data and convert them to data frames. I'll stick with the top 20 aggregate data. We also want to get the mean mileage of all the top 20 brands and make them into a single data frame.

In [44]:
mileage_data ={}
for b in top_20_brands: 
    selected_rows = autos[autos["brand"] == b]
    mean_mileage = selected_rows["odometer_km"].mean() * .6213
    mileage_data[b] = mean_mileage

In [45]:
mileage_series = pd.Series(mileage_data)

In [46]:
price_series = pd.Series(aggregate_data)

In [47]:
brands_mean_df = pd.DataFrame(mileage_series, columns=["mean_mileage"])

In [48]:
brands_mean_df["mean_price"] = price_series

In [49]:
brands_mean_df

mean_mileage    mean_price
volkswagen      79969.780841   6729.819564
bmw             82369.409790   8571.480148
opel            80342.884432   5432.479196
mercedes_benz   81258.810127   8628.450366
audi            80245.484410   9336.687454
ford            77192.481317   7456.547933
renault         79570.718083   2474.864607
peugeot         79000.547380   3094.017229
fiat            72767.880952   2813.748538
seat            75258.877491   4397.230950
skoda           68870.212794   6368.000000
nissan          73519.047686   4743.402525
mazda           77329.504231   4112.596615
smart           61711.726929   3580.223903
citroen         74290.406107  45486.685496
toyota          72036.225126   5167.091062
hyundai         66132.605769   5365.254274
sonstige_autos  55735.057484  45208.364425
volvo           85781.596019   4946.501171
mini            54739.720049  10613.459658

It seems that on average regardles of mileage the closer the cars are to the top the more expnsive they'll be.

# 9. Next Steps

In this guided project, we practiced applying a variety of pandas methods to explore and understand a data set on car listings. Here are some next steps for you to consider:

Data cleaning next steps:
Identify categorical data that uses german words, translate them and map the values to their english counterparts
Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
See if there are particular keywords in the name column that you can extract as new columns
Analysis next steps:
Find the most common brand/model combinations
Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
How much cheaper are cars with damage than their non-damaged counterparts?